Added whitelist feature

glubsy opened the pull request:
#67 Easy enough.

mrichar1 commented:
Thanks for suggesting these changes!

I'd be tempted to try to make this a bit more DRY - repeated lines of code usually mean there's a better way to do it! There's also the issue of what happens if something is in both lists - precedence needs to be decided upon and explicitly documented - which probably also involves merging the code a bit more.

It probably also makes sense for the config option names to line up - i.e have whitelist and blacklist - otherwise it's potentially confusing.

glubsy commented:
Regarding the issue of an item being in both lists at the same time, here in its current state, it's fairly straightforward: if an app is in both whitelist and blacklist, the blacklist being checked after the whitelist, the app is blocked. No risk of copying from a security-sensitive program by mistake that way

I guess you could say the blacklist has precedence over whitelist for security reasons. ;)

I agree perhaps it needs to be documented a bit better. I forgot that the clipster.ini is not written to disk by with default values on first start of clipster. But this is what I added to my clipster.ini for example:

```python
# Comma-separated list of WM_CLASS properties for apps where clipboard changes should be ignored.
# Used to ignore clipboard changes from sensitive apps, e.g. password managers.
blacklist_classes = chromium,subl3

# Comma-separated list of WM_CLASS properties for apps where clipboard changes should NOT be ignored.
# Used to only monitor clipboard changes from whitelisted apps, all other apps will be ignored!
whitelist_classes = chromium
```

I'll update this branch with the changes.

glubsy commented:
If you're up for it, I could look for more of these small optimizations. #Meltdown kernel patches slowing down CPUs

glubsy commented:
There is a bug currently with setting the whitelist_classes to '', since it counts for a non empty value thus blocking capturing anything. Will fix.

glubsy commented on clipster:
```diff
@@ -458,7 +462,7 @@ class Daemon(object):
 
         # Only monitor owner-change events for apps with certain WM_CLASS values
-        if self.whitelist_classes:
+        if not isinstance(self.whitelist_classes, str):
```
I don't really like this, because the type of whitelist_classes can either be string if empty, or list if non empty. But it works. What do you think?

mrichar1 commented:
> There is a bug currently with setting the whitelist_classes to ''. It counts as a non empty value due to the .lower().split(',') methods returning an empty string in the list,

So I realised we don't need a default value/fallback, since one is provided in the `config_defaults` dict in `parse_config()`

However I can't replicate this at all:

```
import configparser
config = configparser.SafeConfigParser({"filter_classes": ""})
config.add_section('clipster')
print(config.get('clipster', 'filter_classes').lower().split())
# []
```

A couple of meta-points...

* `fallback` is python3 only - so be careful to check such features if adding them.

* It's fine to have opened multiple PRs for the separate issues - it actually makes it slightly easier in the general case - it keeps things more atomic, it helps for example if 2 different devs each focus on a particular aspect, and it makes it easier to only accept half of a set of changes if they are completely separate. Ideally this would be done as first implementing (or not) the optimization changes, and then adding the new white/blacklist stuff, but it's a small enough set just now it's fine to keep it all going in this PR :smiley:

glubsy commented:
Thanks!

Regarding the double quote issue, I forgot to mention that more in details yesterday (was working too late). This is reproducible by writing `whitelist_classes = ""` instead of `whitelist_classes = ` in the config file, which was the case before thus could trigger this bug for users who didn't change their config files after this change. 

I've made the code a bit more "dry" by making a function to split the option string. It's not used anywhere else though, maybe it wasn't worth it...?

mrichar1 commented:
That looks a bit nicer - I hadn't appreciated that in python3 quotes in the inifile are treated as a literal string!

I'd be tempted to make (and maybe rename?) `parse_option_string` always returns a list, either empty, or containing the resulting split. That way we can do away with the `isinstance` and go back to just testing the truthyness of it. Also `if string != ''` is equivalent to `if string`.

So something like:

```
if string and string != r'""':
    return string.lower().split(',')
return []
```

Also probably don't need to make the `*_classes` methods attrs of `self` - they're only used in this method, so no need to make them available to other parts of the class. (Of course, if they are to be needed elsewhere then they can stay in `self`, but it's a good idea to create them empty in `__init__()` to make it clear to anyone reading the code that they might exist in the future).

mrichar1 commented:
Sorry - ignore that comment - I'm looking at a different project as well just now and got mixed up as to what was happening where! 

Yes, you're quite right that variables are instantiated in `__init__` (as `self` attrs) and then used in `owner_change` repeatedly. Sorry to make you go to the effort of those parts of 7c10715 !

(This makes me wonder if there are other optimisations that can happen elsewhere through the use of `@lru_cache` decorators on methods, but then that's a different story).

mrichar1 commented:
The other thing I mentioned earlier about making the code more DRY - I think you can collapse it down a lot with a single logic test, something like:

```
# Wnck setup stuff here...
class_name = wm_class.lower()
if (self.whitelist_classes and class_name not in self.whitelist_classes) or if self.blacklist_classes and class_name in self.blacklist_classes):
    return True
```
 Obviously the order of the logic is important for setting correct precedence for the white/blacklist.

glubsy commented:
Thanks! Alright then. I've reverted that back to parsing the whitelist and blacklist during init(). 

Regarding collapsing logic tests, I don't think it's possible to do the test in one swoop without getting the active window's WM_CLASS every time first, regardless of whether the *lists are empty or not. That's not very efficient. I think we want to ignore the WM_CLASS if both lists are empty.

In this new commit, we can do it in one pass. 

Regarding the decorator solution, I'm not familiar enough with decorators yet so I'm not sure how to do that, but there are definitely other places where the parsing of config dictionaries could be done only once and not so often. For example, in the very same owner_change() method: 
` active = self.config.get('clipster', 'active_selections').split(',')` could be done during the Daemon's `init()` as well.

mrichar1 commented:
Yep - that implementation is exactly what I was meaning (sorry if my short example wasn't clear!) - obviously the active window stuff doesn't have to be in it's own method, but it makes it a bit clearer and easier to test if it is, so I approve.

Now I'm looking at the code a bit more, there's another potential improvement in this area. Since the whole classes filtering stuff relies on Wnck working, perhaps it is worthwhile moving the logic that tests for Wnck into `__init__`, and having something like:

```
whitelist_classes = blacklist_classes = []
if Wnck:
    self.blacklist_classes = get_list_from_option_string(self.config.get('clipster', 'blacklist_classes'))
    self.whitelist_classes = get_list_from_option_string(self.config.get('clipster', 'whitelist_classes'))
else:
    logging.error("'whitelist_classes' or 'blacklist_classes' require Wnck (libwnck3).")
```

That way `owner_change` doesn't have to test `Wnck` on each iteration. What do you think?

glubsy commented:
Thank. Yes! I thought about testing for Wnck in init() as well. That's even better that way!

And yes you are right, get_wm_class_from_active_window() is a bit unnecessary now that the two lists comparisons are done in one pass. I forgot about that after the changes, but it does make is a bit easier to read. 

Looks better now.

mrichar1 commented:
That looks great now - am happy to merge. If you're still bored and wanting more to do then you could always update the test suite to test `get_list_from_option_string()` but it's a simple enough method I won't reject the PR without one :wink:  Future optimisations also always welcome for review!

I'll hopefully have a look #66 at the start of next week. Thanks!

glubsy commented:
Thanks! Sorry for being messy and having you so involved in correcting my mistakes. I'm still learning a lot. ;-)

Speaking of which, I always wanted to become more familiar with unit testing, so I'll take the opportunity to patch the test suite and do a new PR. 

I'll have another look at some more optimizations and propose another PR. Cheers~

mrichar1 commented:
No problem - always happy to help others learn - look forward to some more PRs :+1:
